compiler/semantic/sql.go

```diff
@@ -478,9 +478,40 @@ func (t *translator) sqlQueryBody(query ast.SQLQueryBody, seq sem.Seq) (sem.Seq,
 		return seq, sch
 	case *ast.SQLUnion:
 		left, leftSch := t.sqlQueryBody(query.Left, seq)
-		left, _ = endScope(query.Left.(ast.Node), leftSch, left)
+		left, leftSch = endScope(query.Left.(ast.Node), leftSch, left)
+		leftStatic, lok := leftSch.(*staticSchema)
+		if !lok {
+			t.error(query.Left, errors.New("set operations cannot be applied to dynamic sources"))
+		}
 		right, rightSch := t.sqlQueryBody(query.Right, seq)
-		right, _ = endScope(query.Right.(ast.Node), rightSch, right)
+		right, rightSch = endScope(query.Right.(ast.Node), rightSch, right)
+		rightStatic, rok := rightSch.(*staticSchema)
+		if !rok {
+			t.error(query.Right, errors.New("set operations cannot be applied to dynamic sources"))
+		}
```

Comment:
Shouldn't this work for any schema that knows its output columns? Any schema implementations that wraps another schema might know that (depending on what it wraps).

Reply:
My thinking here was that after the `endScope` call the `schema` should be either static or dynamic- I had tested this with selects but not with joins schemas. Turns out this is correct but there's an issue with select/joins on a query like this:
```
select *
from (values {x:1})
inner join (values {y:1}) on x = y
union all
select * from (values {y:2,z:2})
```
that has the left side returning dynamic and so we errantly get the error message:
```
set operations cannot be applied to dynamic sources
```

I'm going to investigate why this is happening.

Reply:
@nwt I fixed this issue here https://github.com/brimdata/super/pull/6360 will rebase this branch on top of this. Used your idea of adding `getColumns` to the schema interface.